### ISSUE TITLE
Native hooking using our library

### ISSUE THREAD

**Issue by pylerSM** [devadvance/rootcloak#62]
LD_PRELOAD is command thing in *nix world. http://www.catonmat.net/blog/simple-ld-preload-tutorial-part-2/
We can use our "system" code and/or real system code then.

Since Android supports wrapping (ie. we can use LD_PRELOAD) we could create library to fake some system calls (fopen, etc). and then just run on rooted device
"setprop wrap.com.package.name LD_PRELOAD=/path/to/lib/librootcloak.so" after every device boot.

Ideas?

**Comment by hikaritenchi** [comment]
I think this is worth at least a proof of concept. 

Are we aware of any kind of potential performance hit? Could this potentially break in newer NDK?

If it does make it into RootCloak, I'd imagine it'd be a toggle.

**Comment by pylerSM** [comment]
´´´
#define _GNU_SOURCE

#include <stdio.h>
#include <dlfcn.h>
#include <android/log.h>


FILE *fopen(const char *path, const char *mode) {
    printf("In our own fopen, opening %s\n", path);
    __android_log_print(ANDROID_LOG_INFO, "ROOTCLOAK", "path %s, mode %s", path, mode);

    FILE *(*original_fopen)(const char*, const char*);
    original_fopen = dlsym(RTLD_NEXT, "fopen");
    return (*original_fopen)(path, mode);
}
´´´

and I use ´´´
  LOCAL_PATH := $(call my-dir)
include $(CLEAR_VARS)

  LOCAL_SRC_FILES := rootcloak.c

  LOCAL_CFLAGS := -fPIC -shared

  LOCAL_LDLIBS := -llog

  LOCAL_MODULE := librootcloak

  include $(BUILD_SHARED_LIBRARY)
´´´

It was built but nothing happened after aj setprop-ed it :D

**Comment by NHellFire** [comment]
Performance hit from this should be basically none.
@pylerSM: Need to use the absolute path for it.
I'm attempting to test it, but it causes root inspector to crash at launch with EOFException

**Comment by hahaopsmeow** [comment]
@NHellFire does it work on Linux box? If it works on Linux, but doesn't work on Android - it means that Android has protection against this approach...

**Comment by pylerSM** [comment]
https://cedricvb.be/post/intercepting-android-native-library-calls/

**Comment by pylerSM** [comment]
https://github.com/apitrace/apitrace/issues/296

**Comment by pylerSM** [comment]
@NHellFire Okey, i use now
setprop wrap.com.devadvance.rootinspect LD_PRELOAD=/data/local/librootcloak.so


but I wasnt able to get even your exception you got.. Maybe you can show what you do?

**Comment by NHellFire** [comment]
@hahaopsmeow: We're not modifying the environment directly, we're telling Android to run the app in a wrapper. Which is documented in the debugging native code Android docs.

@pylerSM: What Android version are you testing on? Maybe it doesn't support wrappers. I'm using rooted stock marshmallow on my S5
I put mine in / to make sure it's readable by everyone and ran: setprop wrap.com.devadvance.rootinspect LD_PRELOAD=/librootcloak.so

I wonder if my problems are caused be SELinux. 

Works as root:
root@klte:/ # LD_PRELOAD=/librootcloak.so date 
In our own fopen, opening /proc/stat
Thu Jul  7 22:01:07 BST 2016

Shell user:
shell@klte:/ $ LD_PRELOAD=/librootcloak.so date
CANNOT LINK EXECUTABLE: couldn't map "/librootcloak.so" segment 2: Permission denied
page record for 0xb6eac04c was not found (block_size=32)

I'll need to check with a custom kernel.

**Comment by devadvance** [comment]
SELinux definitely interferes. I encountered this when doing RootCloak Plus
(cydia-based). You have to turn off enforcement.

**Comment by pylerSM** [comment]
NHellFire , I tested it on Genymotion 6.0 device. But I will look on it...

**Comment by pylerSM** [comment]
NHellFire , I tested it on Genymotion 6.0 device. But I will look on it...

Edit:
LD_PRELOAD=./data/app/com.devadvance.rootcloak2.debug-1/lib/x86/librootcloak.so date seems to work fine

**Comment by pylerSM** [comment]
setprop wrap.com.devadvance.rootinsp LD_PRELOAD=/data/app/com.devadvance.rootcloak2.debug-1/lib/x86/librootcloak.so

setprop wrap.com.pyler.myapp LD_PRELOAD=/data/app/com.devadvance.rootcloak2.debug-1/lib/x86/librootcloak.so

**Comment by pylerSM** [comment]
if I tried to run
setprop wrap.com.devadvance.rootinsp LD_PRELOAD=/data/app/com.devadvance.rootcloak2.debug-1/lib/x86/librootcloak.so and they I launched RootInspector. I get no messgae in log, even no " Error reading pid from wrapped process, child may have died"

**Comment by Phoenix09** [comment]
I tried with root inspector on my S6 with permissive kernel, I get the same exception @nhellfire does

**Comment by NHellFire** [comment]
I'll do some testing in the emulator when I've got time. Android is supposed to still allow it as it's needed for debugging memory leaks in native apps. If I can get my old phone working again, I'll try in CM too.. maybe Samsung patched it to not allow it.

**Comment by pylerSM** [comment]
Good :) I think I will also hack it on my old Samsung phone, emulator can have undefined behaviour in this case.

**Comment by pylerSM** [comment]
`findAndHookMethod("java.lang.Runtime", lpparam.classLoader, "loadLibrary", String.class, ClassLoader.class, new XC_MethodReplacement() {
                    @Override
                    protected Object replaceHookedMethod(MethodHookParam param) throws Throwable {
                        String libname = (String) param.args[0];

                        String soLibrary = "/data/lib" + libname + ".so";
                        File patchedLibrary = new File(soLibrary);
                        //
                        if (patchedLibrary.exists()) {
                            System.load("/data/libHelloWorld.so");
                            XposedBridge.log("called");
                        } else {
                            XposedBridge.invokeOriginalMethod(param.method, param.thisObject,  param.args);

                        }
                        return null;
                    }
                }
        );`

**Comment by Phoenix09** [comment]
ok, Root Inspector with that library works as expected on my Nexus 7 with CM13
but on my S6 and @NHellFire's S5 it crashes, I wonder if Samsung intentionally broke it

**Comment by NHellFire** [comment]
For the emulator, I had to use this, otherwise I got the same error as my S5.

/data/local/rootcloak-wrapper.sh (make sure to chmod 755):
```
#!/system/bin/sh
export LD_PRELOAD=/data/local/librootcloak.so
exec $*
```

And setprop: `setprop wrap.com.blah "logwrapper /data/local/rootcloak-wrapper.sh"`

**Comment by pylerSM** [comment]
but when you do this and setprop wrap. (rootinspector package) "logwrapper /data/local/rootcloak-wrapper.sh"

have you any logs in logcat with "path %s mode %s"?

I tested it on Genymotion Nexus 6P 6.0 and still no luck, no logs in logcat.

**Comment by NHellFire** [comment]
I tested a game to see how it detects xposed and got a lots of logged fopen
calls, yes. Other functions that'd need to be wrapped are stat() as some
apps may just use that to check for existence

**Comment by pylerSM** [comment]
I think RootCloakPlus lib can be easily ported https://github.com/devadvance/rootcloakplus/blob/master/jni/rootcloakplus.cy.cpp now.

**Comment by pylerSM** [comment]
rootcloak-wrapper.sh and librootcloak.so to /data/local... chmod them to 755. setproped.

Then I launched root inspector and I am not able to see any logs.

Nexus 7 2012 and Android 5.1...

If it would work, I think https://github.com/devadvance/rootcloakplus/blob/master/jni/rootcloakplus.cy.cpp needs small rework only...

**Comment by devadvance** [comment]
Most of the code from RootCloak Plus is fairly portable, though there might be some work that needs to be done for how I filter reading magic files. That was an interesting bit of interception.

**Comment by NHellFire** [comment]
Working for me in the emulator:
```
130|root@generic_x86_64:/ # logcat -c; logcat | grep ROOTCLOAK
07-14 19:49:30.991  3350  3375 I ROOTCLOAK: path /system/xbin/su, mode r
07-14 19:49:31.005  3377  3377 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:31.024  3378  3378 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:31.073  3379  3379 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:31.108  3380  3380 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:31.127  3382  3382 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:31.137  3383  3383 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:31.148  3384  3384 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:31.790  3384  3384 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:31.807  3384  3384 I ROOTCLOAK: path /proc/self/task/3384/maps, mode re
07-14 19:49:31.944  3393  3393 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:32.647  3393  3393 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:32.692  3393  3393 I ROOTCLOAK: path /proc/self/task/3393/maps, mode re
07-14 19:49:33.006  3403  3403 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:33.175  3403  3403 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:33.186  3403  3403 I ROOTCLOAK: path /proc/self/task/3403/maps, mode re
07-14 19:49:33.250  3412  3412 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:33.417  3412  3412 I ROOTCLOAK: path /proc/stat, mode re
07-14 19:49:33.424  3412  3412 I ROOTCLOAK: path /proc/self/task/3412/maps, mode re
07-14 19:49:33.483  3350  3375 I ROOTCLOAK: path /system/xbin/su, mode r
```

I just put lib and shell script in /data/local/ and chmod 755 both:
```
130|root@generic_x86_64:/ # getprop | grep wrap
[wrap.com.devadvance.rootinspect]: [logwrapper /data/local/rootcloak-wrapper.sh]
root@generic_x86_64:/ # ls -ll /data/local/*rootcloak*
-rwxr-xr-x root     root         5936 2016-07-10 15:40 librootcloak.so
-rwxr-xr-x root     root           71 2016-07-10 23:16 rootcloak-wrapper.sh
```

**Comment by NHellFire** [comment]
Here's my current source I'm testing:
```
#define _GNU_SOURCE

// stat
#include <libgen.h>
#include <sys/stat.h>
#include <string.h>

// fopen
#include <stdio.h>

// readdir
#include <dirent.h>

// Required by all
#include <dlfcn.h>
#include <android/log.h>
#include <errno.h>


FILE *fopen(const char *path, const char *mode) {
	printf("In our own fopen, opening %s\n", path);
	__android_log_print(ANDROID_LOG_INFO, "ROOTCLOAK", "fopen(): path %s, mode %s", path, mode);

	char *fname = basename(path);

	if (strcasecmp("su", fname) == 0 || strcasecmp("daemonsu", fname) == 0 || strcasecmp("superuser.apk", fname) == 0) {
		__android_log_print(ANDROID_LOG_INFO, "ROOTCLOAK", "fopen(): Hiding su file %s", path);
		errno = ENOENT;
		return NULL;
	}

	static FILE *(*original_fopen)(const char*, const char*) = NULL;
	if (!original_fopen) {
		original_fopen = dlsym(RTLD_NEXT, "fopen");
	}
	return original_fopen(path, mode);
}


int stat(const char *path, struct stat *buf) {
	printf("In our own stat, stat()'ing %s\n", path);
	__android_log_print(ANDROID_LOG_INFO, "ROOTCLOAK", "stat(): path %s", path);

	char *fname = basename(path);

	if (strcasecmp("su", fname) == 0 || strcasecmp("daemonsu", fname) == 0 || strcasecmp("superuser.apk", fname) == 0) {
		__android_log_print(ANDROID_LOG_INFO, "ROOTCLOAK", "stat(): Hiding su file %s", path);
		errno = ENOENT;
		return -1;
	}


	static int (*original_stat)(const char*, struct stat*) = NULL;
	if (!original_stat) {
		original_stat = dlsym(RTLD_NEXT, "stat");
	}
	return (int) original_stat(path, buf);
}

int lstat(const char *path, struct stat *buf) {
	printf("In our own lstat, lstat()'ing %s\n", path);
	__android_log_print(ANDROID_LOG_INFO, "ROOTCLOAK", "stat(): path %s", path);

	char *fname = basename(path);

	if (strcasecmp("su", fname) == 0 || strcasecmp("daemonsu", fname) == 0 || strcasecmp("superuser.apk", fname) == 0) {
		__android_log_print(ANDROID_LOG_INFO, "ROOTCLOAK", "stat(): Hiding su file %s", path);
		errno = ENOENT;
		return -1;
	}


	static int (*original_lstat)(const char*, struct stat*) = NULL;
	if (!original_lstat) {
		original_lstat = dlsym(RTLD_NEXT, "lstat");
	}
	return (int) original_lstat(path, buf);
}

struct dirent *readdir(DIR *dirp) {
	printf("In our own readdir\n");
	__android_log_print(ANDROID_LOG_INFO, "ROOTCLOAK", "readdir()");

	static struct dirent *(*original_readdir)(DIR*);
	if (!original_readdir) {
		original_readdir = dlsym(RTLD_NEXT, "readdir");
	}

	struct dirent* ret = original_readdir(dirp);
	if (ret == NULL) {
		return ret;
	}

	printf("readdir(): d_name = %s\n", ret->d_name);
	__android_log_print(ANDROID_LOG_INFO, "ROOTCLOAK", "readdir(): d_name = %s", ret->d_name);

	unsigned int found = 0;
	do {
		if (strcasecmp("su", ret->d_name) == 0 || strcasecmp("daemonsu", ret->d_name) == 0 || strcasecmp("superuser.apk", ret->d_name) == 0) {
			printf("Found su file, reading next...");
			ret = original_readdir(dirp);
			printf(" done!\n");
		} else {
			found = 0;
		}
	} while (found == 1 && ret != NULL);


	return ret;
}
```

I've patched various functions that are used for checking file existence, including readdir() for ls. With this version, root inspector's native checks only report found for `pm list`. I've also tidied the code a little, including making original_blah static so that dlsym() only needs to be called once, rather than for each call.

```
root@generic_x86_64:/ # ls -ll /system/xbin/su
-rwsr-xr-x root     shell       10280 2016-05-16 20:04 su
root@generic_x86_64:/ # LD_PRELOAD=/data/local/librootcloak.so ls -ll /system/xbin/su                     <
In our own fopen, opening /proc/stat
In our own lstat, lstat()'ing /system/xbin/su
/system/xbin/su: No such file or directory
```

**Comment by pylerSM** [comment]
execlp("ls", "ls", path, (char *)0);
execlp("pm", "pm", "list", "packages", (char *)0);

so I believe execlp should be "hooked"

**Comment by pylerSM** [comment]
Oh, but now I get Error reading pid error:
http://pastebin.com/NusmhKJv

"executing /data/local/rootcloak-wrapper.sh failed: No such file or directory"

but I have it
130|root@vbox86p:/ # ls -ll /data/local
-rwxr-xr-x root     root         9508 2016-07-15 05:32 librootcloak.so
-rwxr-xr-x root     root           72 2016-07-12 04:56 rootcloak-wrapper.sh

**Comment by pylerSM** [comment]
same code, just compiled for ARM gave me:

root@tilapia:/data/app/com.devadvance.rootcloak2.debug-2/lib/arm # LD_PRELOAD=./librootcloak.so date
WARNING: linker: librootcloak.so: unused DT entry: type 0x6ffffffe arg 0xcf4
WARNING: linker: librootcloak.so: unused DT entry: type 0x6fffffff arg 0x2

**Comment by pylerSM** [comment]
Maybe because system API cant see "/data/local"?

**Comment by Phoenix09** [comment]
@pylerSM, try creating a world readable directory in /data

also, `pm` is Java, does rootcloak hide packages from `pm list packages`?
replacing `packages` with something else may not work because the app is expecting packages, it may crash or just refuse to work

**Comment by pylerSM** [comment]
https://android.googlesource.com/platform/frameworks/base/+/android-6.0.0_r3/cmds/pm/src/com/android/commands/pm/Pm.java#330

So, if RootCloaks hooks Android API to get Installed Apps, it should work without needing to touch execlp. Just checker Binder.getCaller API to see if SHELL/ROOT and then fake app list.


my /data
drw-rw-rw- root     root              2000-01-03 04:43 local
drwxrwx--- root     root              1970-01-01 01:00 lost+found

**Comment by NHellFire** [comment]
Yeah, I think hooking getInstalledPackages() from java would be better than patching exec

https://github.com/NHellFire/librootcloak there's the full source I'm working with.

**Comment by pylerSM** [comment]
https://github.com/devadvance/rootcloak/pull/69

Well, but I am still unable to make it working.. i used /data/local, /cache/, /tmp .. I chmoded everything to 777 for sure, lol. still nothinh.
"executing /(location)/rootcloak-wrapper.sh failed: No such file or directory"

**Comment by Phoenix09** [comment]
@pylerSM, is your path correct?
is the path to sh in the script correct?
does it work if you run it manually?
`/data/rootcloak-wrapper.sh ls` for example
if it couldn't access it you should be getting permission denied or
similar, not no such file or directory

On 15 Jul 2016 16:03, "pyler" <notifications@github.com> wrote:

#69 <https://github.com/devadvance/rootcloak/pull/69>

Well, but I am still unable to make it working.. i used /data/local,
/cache/, /tmp .. I chmoded everything to 777 for sure, lol. still nothinh.
"executing /(location)/rootcloak-wrapper.sh failed: No such file or
directory"

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/devadvance/rootcloak/issues/62#issuecomment-232976569>,
or mute the thread

**Comment by pylerSM** [comment]
1|root@vbox86p:/data/local # ls
librootcloak.so
rootcloak-wrapper.sh
tmp
root@vbox86p:/data/local # `/data/local/rootcloak-wrapper.sh ls`
/system/bin/sh: /data/local/rootcloak-wrapper.sh: No such file or directory

ehmmmmm :D

**Comment by devadvance** [comment]
@NHellFire  are you suggesting not patching exec at all? Or just for pm?

**Comment by Phoenix09** [comment]
@pylerSM, check your script :P
alternatively, look at @NHellFire's repo

@devadvance, `pm list packages` use the normal Android API, so no need to modify exec
so if it's run by a library in an app, it will have the same UID as the app so will be easy enough to do in the rootcloak module

**Comment by pylerSM** [comment]
pm list packages use the normal Android API, so no need to modify exec
so if it's run by a library in an app, it will have the same UID as the app so will be easy enough to do in the rootcloak module


ahh, yes!

check your script :P

downloaded file from @NHellFire repo. Now works. WTF :D

**Comment by pylerSM** [comment]
And https://github.com/NHellFire/librootcloak/pull/1 we patched all native calls.

**Comment by pylerSM** [comment]
Please test https://github.com/devadvance/rootcloak/pull/70

**Comment by hikaritenchi** [comment]
@pylerSM is there a pragmatic way of checking this on a physical device?

**Comment by pylerSM** [comment]
Compile test build with my patch, install devadvance's root inspector. Then run RootCloak-Native App Hooking, install lib on first run, and enable RootInspector in provided setting. Then run RootInspector and Native checks page. Everything should be green except pm list packages (this is Java side)

**Comment by pylerSM** [comment]
I would like to add switch for debug logs in native library to avoid logcat flooding. Now I am gonna sleep. :D tomorrow I would do it.

**Comment by pylerSM** [comment]
X86 works. But on ARM I got weird issue like:
WARNING: linker: librootcloak.so: unused DT entry: type 0x6ffffffe arg 0xcf4
WARNING: linker: librootcloak.so: unused DT entry: type 0x6fffffff arg 0x2

solution?

**Comment by devadvance** [comment]
Merged #70 and #72 into the 3.0beta branch. @pylerSM it looks good!

There are additional functions that Root Inspector isn't checking. I can work on migrating those functions from RootCloak Plus (e.g. access, fstat, etc.) and possibly switching out execl for execve (all exec* functions are actually "front-ends" for execve, see http://linux.die.net/man/3/execl).

Does this cover everything from apps like RootBeer as well? https://github.com/scottyab/rootbeer

For reference: https://github.com/devadvance/rootcloakplus/blob/master/jni/rootcloakplus.cy.cpp

**Comment by pylerSM** [comment]
Rootbear native just check for su paths using fopen - fine, we did it.

Port rootcloak plus library would be best, I think.

**Comment by pylerSM** [comment]
fstat is not needed, I believe. convert fd to filename, meh :dancer: 
https://github.com/devadvance/rootcloak/pull/73

**Comment by NHellFire** [comment]
Yeah, fstat isn't needed. File would first have to be opened by open/fopen (or similar).. and we're just patching them to return "does not exist".

**Comment by artas182x** [comment]
Hey, recently i was discovering this issue. Google during signing up or using Android Pay uses DroidGuard which checks many strange things like mac address from /sys/class/net/ patch. What's more it also downloads small apk, which is stored on memory in less than one second so nobody can see it. It's called com.google.ccc.abuse.droidguard.droidguasso. I'm attaching droidguard cache which Android stores in /data/data/com.google.android.gms/app_dg_cache/. However, i wasn't able to catch com.google.ccc.abuse.droidguard.droidguasso, however i know it exists for a short time. Maybe you could analyse it with your hooking tools. [``](http://artas182x.tk/DroidGuard.zip)

**Comment by pylerSM** [comment]
@devadvance  v3.0-beta_20160724_1 was incorrectly compiled?
I installed it (x86) and
root@vbox86p:/ # cd /data/local
LD_PRELOAD=./librootcloak.so  date
Segmentation fault

**Comment by pylerSM** [comment]
regcomp causes it since I comment it and it started working.

**Comment by pylerSM** [comment]
temporary workaround is just comment it for now, see patch https://github.com/devadvance/rootcloak/pull/88 and re-release beta.

@NHellFire can you check it for possible future fix?

**Comment by pylerSM** [comment]
@devadvance also I added auto library updater to update binary in /data/local everytime RootCloak app is updated. Handy I think.

https://github.com/devadvance/rootcloak/pull/87

**Comment by NHellFire** [comment]
Hmm, it worked fine for me. I'll do some testing, see if I can reproduce it

**Comment by hahaopsmeow** [comment]
Hi guys...
Some not good news are here: http://forum.xda-developers.com/xposed/unofficial-systemless-xposed-t3388268/page133

In short now SNED detects systemless root and if it is there it fails...

**Comment by pylerSM** [comment]
Sure, but until we dont know what it checks, we cant bypass it.

**Comment by hahaopsmeow** [comment]
@pylerSM that is true. With systemless root boot.img is modified. It could be that it checks for it?

**Comment by pylerSM** [comment]
They can check everything.. it is very easy to see where new implementation goes and then check for new files.

Sadly, not many people work on this.

**Comment by hahaopsmeow** [comment]
True. When I was looking on how to detect Xposed installed I instantly got
6 working methods in Java. I didn't start with have armor... There will be
more...

**Comment by pylerSM** [comment]
I even know a few lines of code which disables any Xposed modification on the app with that code.

**Comment by pylerSM** [comment]
https://code.google.com/p/android/issues/detail?id=93752

Shit, looks like Lollipop has wraping screwed up.. Now I know why I could make it working on JB, Kitkat, MM (emulator, but has it disabled selinux?) but but not on LP.

**Comment by pylerSM** [comment]
Maybe we could hook (not sure if possible even)

WrapperInit main()

beforeHookedMethod
- disable selinux

afterHookedMethod
- enable selinux
 
or hook onCreate and then re-enable selinux.


Currently I will simple to fix it with just selinux disabling. https://github.com/devadvance/rootcloak/pull/92

**Comment by pylerSM** [comment]
Instead of disabling SELinux, we could use supolicy to use specific supolicy to enable wrapping and keep SELinux enforced. 
https://su.chainfire.eu/#selinux-policies


Edit:Okey, I have neat way! No SELinux disabling!

**Comment by pylerSM** [comment]
https://github.com/devadvance/rootcloak/pull/93

@devadvance  please merge and re-release beta :) thanks

**Comment by NHellFire** [comment]
@pylerSM So far I'm unable to get it to crash with that code. It worked in the emulator and on my Desire HD.

**Comment by pylerSM** [comment]
I will re-test it.

**Comment by pylerSM** [comment]
Sadly, still not work. Genymotion Nexus 6P Android 6.0 still seg fault.

**Comment by pylerSM** [comment]
Latest NDK and build tools 23... Please build APK for me to check even this.

**Comment by hahaopsmeow** [comment]
I have Nexus 6P and Nexus 5X and I can test... tell me how...

**Comment by pylerSM** [comment]
Download and Install RootInspector (Inspect via Native Code screen): https://github.com/devadvance/rootinspector/releases/download/v1.0/com.devadvance.rootinspector-v1.0.apk

Steps:
Download rootcloak app, open, go Native Root Detection -> Install library -> Remove native root detection -> Select RootInspector. Close app and open Root Inspector and Inspect via Native Code. Items should be green.


Two versions to test, regex off version works for me, regex on not. Test please :)
[apk.zip](https://github.com/devadvance/rootcloak/files/387985/apk.zip)

**Comment by hahaopsmeow** [comment]
@pylerSM I just flashed stock ROMs on both mobiles.. What should I install first to test? I will do it right now...

**Comment by pylerSM** [comment]
Test both APK .

**Comment by hahaopsmeow** [comment]
@pylerSM So, I will install both APK to untouched stock ROMs. Right?

**Comment by pylerSM** [comment]
Test first one, uninstall, test second one.

**Comment by pylerSM** [comment]
Yes, root needed.

**Comment by hahaopsmeow** [comment]
@pylerSM systemless CF-Auto-Root is enough?

**Comment by pylerSM** [comment]
Yes

**Comment by hahaopsmeow** [comment]
@pylerSM I do test right now on Nexus 5X

**Comment by hahaopsmeow** [comment]
@pylerSM I just tried with `regex on`.

I did everything as you said and native detection goes all green.

note: when I tried native detection before running rootcloak it hunged up somehow. should I tty `regex off`

**Comment by pylerSM** [comment]
interesting, please regex off too

**Comment by pylerSM** [comment]
okey, seems like it (regex filtering) crashes only on emulator

**Comment by hahaopsmeow** [comment]
@pylerSM please check you mailbox

**Comment by hahaopsmeow** [comment]
@pylerSM both versions work.

**Comment by pylerSM** [comment]
Well, fine, so issue is Genymotion emulator :dancer: 

So I sent PR to enable it back on ARM devices https://github.com/devadvance/rootcloak/pull/95

**Comment by hahaopsmeow** [comment]
@pylerSM but java does detects root :-)

**Comment by pylerSM** [comment]
Yes, because you need enable Xposed module in Xposed Installer. :)

**Comment by NHellFire** [comment]
I don't think it's worth disabling those checks for x86 devices just because it crashes in one emulator. I've tested AMIDuOS x86, Google's x86_64 and a Galaxy Tab 3 10.1 (x86) and it doesn't crash. Build tools 23.0.3 and NDK 11c.

**Comment by pylerSM** [comment]
Please test the specific one - Genymontion Nexus 6P 6.0. Emulator is good for development, so if it is broken on emulator. it makes everything slower.

**Comment by NHellFire** [comment]
It isn't broken on other emulators, or actual x86 devices.

**Comment by pylerSM** [comment]
Okey, let it be
https://github.com/devadvance/rootcloak/pull/103

**Comment by pylerSM** [comment]
@devadvance time for new beta?
